### PR TITLE
Mix jmap imap luc

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/pom.xml [hunk 1]
```diff
@@ -639,6 +646,12 @@
             </dependency>
             <dependency>
                 <groupId>org.apache.james</groupId>
+                <artifactId>james-server-mailets-integration-testing</artifactId>
```
Comment (chibenwa) [inline server/pom.xml]: Can be removed ?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/pom.xml [hunk 2]
```diff
@@ -184,6 +184,12 @@
                     <scope>test</scope>
                 </dependency>
                 <dependency>
+                    <groupId>org.apache.james</groupId>
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/pom.xml]: Is that still needed ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+@ScenarioScoped
+public class ImapStepdefs {
+
+    private final UserStepdefs userStepdefs;
+    private final Map<String, IMAPMessageReader> imapConnections;
+
+    @Inject
+    private ImapStepdefs(UserStepdefs userStepdefs) {
+        this.userStepdefs = userStepdefs;
+        this.imapConnections = new HashMap<>();
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Quynh's remark : Maps.newHashMap()

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Then("^the user has a IMAP message in mailbox \"([^\"]*)\"$")
+    public void hasMessageInMailbox(String mailbox) throws Throwable {
+        try (IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Quynh's remark : extract constants

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                imapMessageReader.userReceivedMessageInMailbox(userStepdefs.lastConnectedUser,
+                    userStepdefs.passwordByUser.get(userStepdefs.lastConnectedUser),
+                    mailbox))
+                .isTrue();
+        }
+    }
+
+    @Then("^the user has a IMAP notification about (\\d+) new message when selecting mailbox \"([^\"]*)\"$")
+    public void hasANotificationAboutNewMessagesInMailbox(int numOfNewMessage, String mailbox) throws Throwable {
+        try (IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Quynh's remark : extract constants

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                imapMessageReader.userGetNotifiedForNewMessagesWhenSelectingMailbox(userStepdefs.lastConnectedUser,
+                    userStepdefs.passwordByUser.get(userStepdefs.lastConnectedUser),
+                    numOfNewMessage, mailbox))
+                .isTrue();
+        }
+    }
+
+    @Then("^the user does not have a IMAP message in mailbox \"([^\"]*)\"$")
+    public void hasNoMessageInMailbox(String mailbox) throws Throwable {
+        try (IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Quynh's remark : extract constants

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                imapMessageReader.userDoesNotReceiveMessageInMailbox(userStepdefs.lastConnectedUser,
+                    userStepdefs.passwordByUser.get(userStepdefs.lastConnectedUser),
+                    mailbox))
+                .isTrue();
+        }
+    }
+
+    @Given("^the user has a open IMAP connexion with mailbox \"([^\"]*)\" selected")
+    public void openImapConnexionAndSelectMailbox(String mailbox) throws Throwable {
+        IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Quynh's remark : extract constants

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            assertThat(
+                imapMessageReader.userDoesNotReceiveMessageInMailbox(userStepdefs.lastConnectedUser,
+                    userStepdefs.passwordByUser.get(userStepdefs.lastConnectedUser),
+                    mailbox))
+                .isTrue();
+        }
+    }
+
+    @Given("^the user has a open IMAP connexion with mailbox \"([^\"]*)\" selected")
+    public void openImapConnexionAndSelectMailbox(String mailbox) throws Throwable {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: s/connexion/connection/ here and the line above

(Quynh's comment)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+        String login = userStepdefs.lastConnectedUser;
+        String password = userStepdefs.passwordByUser.get(login);
+
+        imapMessageReader.connectAndSelect(login, password, mailbox);
+        imapConnections.put(mailbox, imapMessageReader);
+    }
+
+    @Then("^the user has a IMAP RECENT and  notification about new message with uid (\\d+) on connexion for mailbox \"([^\"]*)\"$")
+    public void checkNotificationForNewMessageOnActiveConnexion(int uid, String mailbox) throws Throwable {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: s/uid/msn/ here and in the given (Raph)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        IMAPMessageReader imapMessageReader = imapConnections.get(mailbox);
+        assertThat(imapMessageReader).isNotNull();
+        assertThat(imapMessageReader.userGetNotifiedForNewMessages(uid)).isTrue();
+    }
+
+    @Then("^the user has IMAP EXPUNGE notification for uid (\\d+) on connexion for mailbox \"([^\"]*)\"$")
+    public void checkExpungeNotificationOnActiveConnexion(int uid, String mailbox) throws Throwable {
+        IMAPMessageReader imapMessageReader = imapConnections.get(mailbox);
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: s/uid/msn/ here and in the given (Raph)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);
+
+        String login = userStepdefs.lastConnectedUser;
+        String password = userStepdefs.passwordByUser.get(login);
+
+        imapMessageReader.connectAndSelect(login, password, mailbox);
+        imapConnections.put(mailbox, imapMessageReader);
+    }
+
+    @Then("^the user has a IMAP RECENT and  notification about new message with uid (\\d+) on connexion for mailbox \"([^\"]*)\"$")
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: You miss a a between **and** and **notification** in the given (Raph)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        try (IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);) {
+            assertThat(
+                imapMessageReader.userDoesNotReceiveMessageInMailbox(userStepdefs.lastConnectedUser,
+                    userStepdefs.passwordByUser.get(userStepdefs.lastConnectedUser),
+                    mailbox))
+                .isTrue();
+        }
+    }
+
+    @Given("^the user has a open IMAP connexion with mailbox \"([^\"]*)\" selected")
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: a**n** open connection (Raph)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]";
+        Request.Post(mainStepdefs.baseUri().setPath("/jmap").build())
+            .addHeader("Authorization", userStepdefs.tokenByUser.get(username).serialize())
+            .bodyString(requestBody, org.apache.http.entity.ContentType.APPLICATION_JSON)
+            .execute()
+            .discardContent();
+        Thread.sleep(1000L);
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java]: This debugging sleep shall be removed

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml [hunk 14]
```diff
@@ -167,6 +167,12 @@
                 </dependency>
                 <dependency>
                     <groupId>org.apache.james</groupId>
+                    <artifactId>james-server-mailets-integration-testing</artifactId>
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml]: Not needed anymore ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java [hunk 15]
```diff
@@ -75,7 +74,6 @@
         this.updatePatchConverter = updatePatchConverter;
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java]: Then couldn't mailboxManager also be removed from the constructor arguments

File: server/testing/pom.xml [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+                  <artifactId>assertj-core</artifactId>
+                  <version>${assertj-3.version}</version>
+                  <scope>test</scope>
+              </dependency>
+              <dependency>
+                  <groupId>org.slf4j</groupId>
+                  <artifactId>slf4j-api</artifactId>
+              </dependency>
+              <dependency>
+                  <groupId>org.testcontainers</groupId>
```
Comment (chibenwa) [inline server/testing/pom.xml]: Test container do not seem needed

File: server/testing/pom.xml [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+                  <scope>test</scope>
+              </dependency>
+              <dependency>
+                  <groupId>ch.qos.logback</groupId>
+                  <artifactId>logback-classic</artifactId>
+                  <version>1.1.6</version>
+                  <scope>test</scope>
+              </dependency>
+              <dependency>
+                  <groupId>com.jayway.jsonpath</groupId>
```
Comment (chibenwa) [inline server/testing/pom.xml]: jsonpath do not seem needed

File: server/testing/pom.xml [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+            </build>
+        </profile>
+        <profile>
+            <id>build-for-jdk-8</id>
+            <activation>
+                <jdk>[1.8,)</jdk>
+            </activation>
+          <dependencies>
+              <dependency>
+                  <groupId>org.apache.james</groupId>
```
Comment (chibenwa) [inline server/testing/pom.xml]: Is this one needed?

File: server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java [hunk 19]
```diff
@@ -68,6 +85,34 @@ private String readFirstMessageInInbox(String user, String password, String para
         return imapClient.getReplyString();
     }
 
+    public boolean userGetNotifiedForNewMessages(int uid) throws IOException {
```
Comment (chibenwa) [inline server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java]: s/uid/msn/

File: server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java [hunk 20]
```diff
@@ -68,6 +85,34 @@ private String readFirstMessageInInbox(String user, String password, String para
...
+            .trimResults()
+            .omitEmptyStrings()
+            .splitToList(replyString);
+        return parts.size() == 3
+            && parts.get(2).contains("OK NOOP completed.")
+            && parts.contains("* " + uid + " EXISTS")
+            && parts.contains("* " + uid + " RECENT");
+    }
+
+    public boolean userGetNotifiedForDeletion(int uid) throws IOException {
```
Comment (chibenwa) [inline server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java]: s/uid/msn/

**Comment by chibenwa** [comment]
Ok, about this all UID / MSN story : 

```
EXISTS response reports the number of messages in the mailbox
```

```
The RECENT response reports the number of messages with the \Recent flag set.
```

but

```
The EXPUNGE response reports that the specified message sequence number has been permanently removed from the mailbox.
```

We need to fix accordingly.

**Comment by chibenwa** [comment]
```
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [INFO] -------------------------------------------------------------
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [ERROR] COMPILATION ERROR : 
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [INFO] -------------------------------------------------------------
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [ERROR] /james-project/server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java:[26,35] package org.apache.commons.net.imap does not exist
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [ERROR] /james-project/server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java:[32,19] cannot find symbol
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846]   symbol:   class IMAPClient
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846]   location: class org.apache.james.utils.IMAPMessageReader
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [ERROR] /james-project/server/testing/src/main/java/org/apache/james/utils/IMAPMessageReader.java:[35,26] cannot find symbol
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846]   symbol:   class IMAPClient
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846]   location: class org.apache.james.utils.IMAPMessageReader
[Building Java 8 mix-jmap-imap-luc-8a3c4588-7ec5-4195-a9bc-f4691d27e846] [INFO] 3 errors 
```

You miss the dependancy to commons-net 3.2

**Comment by rouazana** [comment]
Yes, exactly what I meant: msn for expunge and number for recent/exists.

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
#587

**Pull request closed without merging**